mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java

```diff
@@ -70,20 +80,49 @@ public void onAvailable(@NonNull Network network) {
     @Override
     public void onCapabilitiesChanged(@NonNull Network network,
                                       @NonNull NetworkCapabilities networkCapabilities) {
-      if (networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
-        if (networkCapabilities.hasCapability(NetworkCapabilities.TRANSPORT_WIFI)) {
-          envoyEngine.onDefaultNetworkChanged(EnvoyNetworkType.WLAN);
-        } else if (networkCapabilities.hasCapability(NetworkCapabilities.TRANSPORT_CELLULAR)) {
-          envoyEngine.onDefaultNetworkChanged(EnvoyNetworkType.WWAN);
-        } else {
-          envoyEngine.onDefaultNetworkChanged(EnvoyNetworkType.GENERIC);
+      // `onCapabilities` is guaranteed to be called immediately after `onAvailable`.
```

Comment:
Only after Build.VERSION_CODES.O according to the documentation. Can we make sure this callback is only registered for version O+?

Reply:
The `NetworkCallback` API was added in API level 21 but the guarantee is only starting with `Build.VERSION_CODES.O`.

Reply:
In versions under O, this logic won't work right? In Chromium this callback is only registered for version R+.

Reply:
Let me remove the comment. I don't think it makes sense since we don't do anything with `onAvailable` at the moment. The `onCapabilitiesChanged` callback itself is available since API 21 (https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onCapabilitiesChanged(android.net.Network,%20android.net.NetworkCapabilities)), which is the minimum SDK version that we use: https://github.com/envoyproxy/envoy/blob/1130fca31d99f9c7ceff7fd0a1d2c4ee8ca25064/mobile/bazel/android_artifacts.bzl#L311

The Chromium code probably does something more that requires a higher API.

Reply:
If onCapabilitiesChanged() is called before onAvailable() in versions under O, the current logic needs to happen under onAvailable(), right?  Otherwise, the DNS refreshment may still fail if the network is not available yet.

SDK 21 is added in version L which was before version O. To avoid the complexity, I would suggest to just support network change after version O.

Reply:
I updated the PR description.

Reply:
I put the old comment back.